### PR TITLE
gh-101819: Isolate _io

### PR DESCRIPTION
**Pull request by erlend-aasland**
Proof-of-concept. We will split this change up in multiple PRs

* Issue: gh-101819

### CODE REVIEW

File: Modules/_io/iobase.c [hunk 1]
```diff
@@ -447,7 +452,8 @@ _PyIOBase_check_writable(PyObject *self, PyObject *args)
         return NULL;
     if (res != Py_True) {
         Py_CLEAR(res);
-        iobase_unsupported("File or stream is not writable.");
+        _PyIO_State *state = find_io_state_by_def(Py_TYPE(self));
```
Comment (kumaraditya303) [inline Modules/_io/iobase.c]: `find_io_state_by_def` cannot be used here, you need to pass down the state from the object itself.

**Review by kumaraditya303 — DISMISSED** [review]

File: Modules/_io/iobase.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_io/iobase.c]: Well observed. The comment above also confirms this.

**Comment by erlend-aasland** [comment]
@kumaraditya303: The test suite now runs without segfaults (but with failures). I need to check code coverage to see if we've got valid state access everywhere, though.

Most importantly, there's now some highly experimental code in here, namely a51823a6ece769f9c3af757b270f34755b401968. Could you take a look at the changes in that particular commit, if you have time?

File: Modules/_io/iobase.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_io/iobase.c]: Hm, my current fix won't work, unfortunately.

Comment (kumaraditya303) [reply Modules/_io/iobase.c]: You need to pass down the state from the calling function here, `self` can be any object so you cannot get the module state from it. 0cdd1eae3d33dc0fcf2ee25c395a81b262876737 also needs to be reverted.

Comment (erlend-aasland) [reply Modules/_io/iobase.c]: Sure, we can pass state as a parameter to the `_PyIOBase_check_` functions for internal `_io` use, but we'll need to remove them from `iobase_methods` as the new argument spec will be incompatible with `METH_NOARGS`.

File: Modules/_io/iobase.c [hunk 2]
```diff
@@ -797,11 +806,6 @@ static PyMethodDef iobase_methods[] = {
     _IO__IOBASE_READABLE_METHODDEF
     _IO__IOBASE_WRITABLE_METHODDEF
 
-    {"_checkClosed",   _PyIOBase_check_closed, METH_NOARGS},
```
Comment (kumaraditya303) [inline Modules/_io/iobase.c]: Maybe you forgot to add new implementations of these, I cannot find it anywhere else.

Comment (erlend-aasland) [reply Modules/_io/iobase.c]: Hm, that's possible, but I don't remember; it's been a while since I looked at this.

**Comment by kumaraditya303** [comment]
FWIW, it seems better to split it now, it will be more manageable to merge.

**Comment by vstinner** [comment]
> I agree. I'll split it up.

I created PR #104178 based on your PR: it "only" gets types from the module state, without actually converting static types to heap types.

**Comment by vstinner** [comment]
I merged to subset of this PR: https://github.com/python/cpython/commit/c84029179c3287f9c357ccac231fe78469c6f068

@erlend-aasland: Do you want to rebase this PR on top of it? If can try to do it as a new PR if you want.

**Comment by erlend-aasland** [comment]
> I merged to subset of this PR: [c840291](https://github.com/python/cpython/commit/c84029179c3287f9c357ccac231fe78469c6f068)
> 
> @erlend-aasland: Do you want to rebase this PR on top of it? If can try to do it as a new PR if you want.

Great, thanks! I'll sync with main and update this PR.

**Comment by erlend-aasland** [comment]
@kumaraditya303, @vstinner: this is getting slimmer and easier to review. There are still refleaks, though. If we can pin those down, we can land this.

Also, I see no more options to break this further up, at the moment. Let me know if you spot something we can separate out.

**Comment by erlend-aasland** [comment]
Also, regarding the pickle fixes, here's a reminder from an old issue: https://github.com/python/cpython/issues/85224#issuecomment-1093875259

**Comment by erlend-aasland** [comment]
Other things to separate out from this:

- a PR that adds `defining_class` to clinic where needed
- a PR for the pickle workaround
- a PR that prepares the check readable/writable/seekable/closed helpers

**Comment by vstinner** [comment]
> a PR for the pickle workaround

I'm now scared!

**Comment by vstinner** [comment]
> Getting closer and closer to a manageable size. Hopefully it will be easier to find the refleaks then!

This PR has 77 commits. Would it be possible to squash all of them to have a single commit?

This PR is still marked as draft. Remove this flag once you expect a review.

Or you can maybe push remaining changes in smaller PRs, I don't know. It's up to you.

**Comment by erlend-aasland** [comment]
> Or you can maybe push remaining changes in smaller PRs, I don't know. It's up to you.

I think I can still split it up, but the remaining changes are more complicated now. I'll put up a PR for porting TextIOBase to heap types. Locally, it leaks some refs. I need more eyes on it.

**Comment by vstinner** [comment]
Quick fix for the _io.RawIOBase leak:

```diff
diff --git a/Modules/_io/iobase.c b/Modules/_io/iobase.c
index 6ce34a1cbf..14a06f74ad 100644
--- a/Modules/_io/iobase.c
+++ b/Modules/_io/iobase.c
@@ -1059,7 +1059,6 @@ static PyMethodDef rawiobase_methods[] = {
 };
 
 static PyType_Slot rawiobase_slots[] = {
-    {Py_tp_dealloc, rawiobase_dealloc},
     {Py_tp_doc, (void *)rawiobase_doc},
     {Py_tp_methods, rawiobase_methods},
     {Py_tp_traverse, rawiobase_traverse},
```

Note: I used ``./python -m test.bisect_cmd -o bisect -R 3:3 test_io  -v`` to identify the leaks.

**Comment by vstinner** [comment]
Ah, _io.TextIOBase dealloc function has the same issue. Quick fix for test_io:

```diff
diff --git a/Modules/_io/iobase.c b/Modules/_io/iobase.c
index 6ce34a1cbf..14a06f74ad 100644
--- a/Modules/_io/iobase.c
+++ b/Modules/_io/iobase.c
@@ -1059,7 +1059,6 @@ static PyMethodDef rawiobase_methods[] = {
 };
 
 static PyType_Slot rawiobase_slots[] = {
-    {Py_tp_dealloc, rawiobase_dealloc},
     {Py_tp_doc, (void *)rawiobase_doc},
     {Py_tp_methods, rawiobase_methods},
     {Py_tp_traverse, rawiobase_traverse},
diff --git a/Modules/_io/textio.c b/Modules/_io/textio.c
index ad8e7c74df..b6d1113f98 100644
--- a/Modules/_io/textio.c
+++ b/Modules/_io/textio.c
@@ -196,7 +196,6 @@ static PyGetSetDef textiobase_getset[] = {
 };
 
 static PyType_Slot textiobase_slots[] = {
-    {Py_tp_dealloc, textiobase_dealloc},
     {Py_tp_doc, (void *)textiobase_doc},
     {Py_tp_methods, textiobase_methods},
     {Py_tp_getset, textiobase_getset},
```

With this patch, ``./python -m test -v -R 3:3 test_io`` no longer leaks.

**Comment by vstinner** [comment]
Oh wait, I sent my comment too quickly! In fact 3 classes have the same dealloc issue:

* _io.RawIOBase
* _io.BufferedIOBase
* _io.TextIOBase

Now with the patch below, ``./python -m test -v -R 3:3 test_io`` no longer leaks for real :-)

```diff
diff --git a/Modules/_io/bufferedio.c b/Modules/_io/bufferedio.c
index 00698883c7..622088628a 100644
--- a/Modules/_io/bufferedio.c
+++ b/Modules/_io/bufferedio.c
@@ -2450,7 +2450,6 @@ static PyMethodDef bufferediobase_methods[] = {
 };
 
 static PyType_Slot bufferediobase_slots[] = {
-    {Py_tp_dealloc, bufferediobase_dealloc},
     {Py_tp_doc, (void *)bufferediobase_doc},
     {Py_tp_methods, bufferediobase_methods},
     {Py_tp_traverse, bufferediobase_traverse},
diff --git a/Modules/_io/iobase.c b/Modules/_io/iobase.c
index 6ce34a1cbf..14a06f74ad 100644
--- a/Modules/_io/iobase.c
+++ b/Modules/_io/iobase.c
@@ -1059,7 +1059,6 @@ static PyMethodDef rawiobase_methods[] = {
 };
 
 static PyType_Slot rawiobase_slots[] = {
-    {Py_tp_dealloc, rawiobase_dealloc},
     {Py_tp_doc, (void *)rawiobase_doc},
     {Py_tp_methods, rawiobase_methods},
     {Py_tp_traverse, rawiobase_traverse},
diff --git a/Modules/_io/textio.c b/Modules/_io/textio.c
index ad8e7c74df..b6d1113f98 100644
--- a/Modules/_io/textio.c
+++ b/Modules/_io/textio.c
@@ -196,7 +196,6 @@ static PyGetSetDef textiobase_getset[] = {
 };
 
 static PyType_Slot textiobase_slots[] = {
-    {Py_tp_dealloc, textiobase_dealloc},
     {Py_tp_doc, (void *)textiobase_doc},
     {Py_tp_methods, textiobase_methods},
     {Py_tp_getset, textiobase_getset},
```

**Comment by vstinner** [comment]
>@kumaraditya303: fixes by Victor

Well, it was a quick fix. My patch doesn't remove the function, so you get warnings like: ‘textiobase_dealloc’ defined but not used [-Wunused-function].

Also, you didn't change the 3rd class: _io.RawIOBase.

**Comment by kumaraditya303** [comment]
> Well, it was a quick fix. My patch doesn't remove the function, so you get warnings like: ‘textiobase_dealloc’ defined but not used [-Wunused-function].

Also, you didn't change the 3rd class: _io.RawIOBase.

Yeah, I have changes locally, checking for refleaks before pushing more changes to this branch.

**Comment by erlend-aasland** [comment]
Thanks for your help, Victor and Kumar! We're landing this today 🎉

**Comment by erlend-aasland** [comment]
Oh, there's a final issue, @kumaraditya303: as @vstinner pointed out, the `__reduce__`/`__reduce_ex__` hacks should not be needed. Can we please resolve that discussion before merging?

**Comment by kumaraditya303** [comment]
> Oh, there's a final issue, vstinner pointed out, the __reduce__/__reduce_ex__ hacks should not be needed. Can we please resolve that discussion before merging? (See https://github.com/python/cpython/pull/104370)

I not sure how it is not needeed. I see similar thing done in `_csv` module.
 https://github.com/python/cpython/blob/b2c1b4da1935639cb89fbbad0ce170a1182537bd/Modules/_csv.c#L536-L549

**Comment by erlend-aasland** [comment]
> I am not sure how it is not needed. I see similar thing done in `_csv` module.

Hm, Serhiy's general fix (gh-22870) was committed just before the \_csv extension module was adapted (gh-23224).

When I try pickling with protocols 0 and 1 in the REPL, I get a correct TypeError for the _io types. It's strange that this is not the behaviour we're also seeing in `test_pickling`.

**Comment by kumaraditya303** [comment]
> When I try pickling with protocols 0 and 1 in the REPL, I get a correct TypeError for the _io types. It's strange that this is not the behaviour we're also seeing in test_pickling.

Did you remove the `_PyIOBase_cannot_pickle` before doing that? I tried and it pickled without error. Anyways I am not really an expert on pickle.

**Comment by erlend-aasland** [comment]
> Did you remove the `_PyIOBase_cannot_pickle` before doing that? I tried and it pickled without error. Anyways I am not really an expert on pickle.

Probably not; I tried again yesterday, and it also pickled without error. I'm not an expert on pickle either, but it seems to me we've found a corner case where Serhiy's fix of 2020 does not apply. One option can be to fix `Lib/copyreg.py`. In any case, I suggest to land this PR with the pickle _workaround_ in it. If we end up patching Serhiy's general pickle fix, we can revert the workaround introduced here (as Serhiy's patch did with other extension modules back in 2020).

**Review by kumaraditya303 — APPROVED** [review]

**Comment by kumaraditya303** [comment]
```console
root@codespaces-62bef3 /w/cpython (isolate-io/poc) [SIGINT]# ./python -m test -R 3:3 test_io
0:00:00 load avg: 7.01 Run tests sequentially
0:00:00 load avg: 7.01 [1/1] test_io
beginning 6 repetitions
123456
......
test_io passed in 3 min 31 sec

== Tests result: SUCCESS ==

1 test OK.

Total duration: 3 min 31 sec
Tests result: SUCCESS

```

**Comment by vstinner** [comment]
It's good to see ``_PyIO_get_module_state()`` going away :-) Also, bye bye my ``static_types`` and ``_PyIO_FiniTypes()`` workarounds!

I'm happy that these changes were merged as a long list of changes (see PR list in issue #101819): if something will go wrong, it will be easier to identify which sub-part of these changes is causing troubles. I expect troubles, but that's fine. We already have to go trough turbulences to modernize Python ;-)

Thanks @erlend-aasland and @kumaraditya303 for your hard work on this complicated _io extension.

**Comment by erlend-aasland** [comment]
@sunmy2019, why did you engage the refleak bots on a closed PR? Did you check the results of the refleak run that Kumar did for the same commit (fe2db1b) yesterday first? (All 31 checks passed.)

**Comment by sunmy2019** [comment]
It should not pass the ref leaks tests. #104510 Just rerun to check what's going on.

Also, I should not trigger builtbot on a merged PR. It seems nothing will be tested. That's my bad.

**Comment by erlend-aasland** [comment]
> Also, I should not trigger builtbot on a merged PR. It seems nothing will be tested. That's my bad.

Yes. And you can examine the refleak run Kumar did post merge. I did yesterday. I'm not sure why that passed, though.

**Comment by vstinner** [comment]
> Yes. And you can examine the refleak run Kumar did post merge. I did yesterday. I'm not sure why that passed, though.

Your PR was created in February. Maybe the code evolved in the meanwhile and  the final "squash + rebase" commit is different than the PR branch ran on buildbots. For such PR which is in the works for a long time, I prefer to manually squash+rebase time to time to workaround this workflow limitation. There are services like https://mergify.com/ which prevent this workflow flaw. I heard that GitHub automerge can do something similar (run GHA jobs on the "final" commit), but buildbots are not handled and use the old way (run the jobs on the PR branch which is not merged into main / rebased).

**Pull request merged**